NO_PARTITION_COLUMNS option to skip partition writes for Parquet copy

ykskb opened the pull request:
Fixes the issue of duplicate columns created for partitioned writes of Parquet files which is mentioned in https://github.com/duckdb/duckdb/issues/12147 and https://github.com/duckdb/duckdb/discussions/11931

This PR provides `NO_PARTITION_COLUMNS` option into Copy-To, which skips the write of partition columns when specified.

ykskb committed:
Add no_parition_columns option for Parquet copy.

Mytherin commented:
Thanks for the PR!

I don't think the Parquet writer is the correct place to fix this. Fixing this only in the Parquet writer means the issue will persist when writing partitioned CSV files or JSON files. Ideally this is fixed generically within the copy to function.

ykskb commented:
Hello @Mytherin, thanks for your feedback!

I did see the possibility of making skip logic in copy function or somewhere before the writers. I wasn't sure if this is required for other file types, so I just wanted to minimize the changes at higher level. Do you mind if I give another try?

Mytherin commented:
Sure, that's fine by me

ykskb committed:
Change exclusion of partition columns more generic.

ykskb committed:
Fix a variable naming.

ykskb committed:
Move up exclusion logic to keep writers intact.

ykskb committed:
Filter types and names early for copy-to.

ykskb committed:
Move column exclusions into physical operator.

ykskb committed:
Fix minor conditions in copy function.

ykskb commented:
Hi @Mytherin, I've moved the exclusion logic of partition columns writes to physical plan & operator. While writer logics remain untouched, the exclusion of partition column writes is working for both CSV and Parquet. (Partitioned writes in JSON format seems unsupported currently.) It'll be great if you can have a look. Thanks!

ykskb committed:
Simplify column exclusion logic handling.

ykskb committed:
Fix a condition for bind call.

ykskb committed:
Fix format.

ykskb committed:
Make function names more descriptive.

ykskb commented:
Simplified the logic and parameters as I realized creating new params in logical/physical operators for partition write skips  was complicating the codes. Now only the exact parameter of `no_partition_columns` option is added. It'll be great if anyone can take a look. Thanks!

Mytherin commented on src/include/duckdb/function/copy_function.hpp:
```diff
@@ -122,6 +122,10 @@ typedef vector<unique_ptr<Expression>> (*copy_to_select_t)(CopyToSelectInput &in
 enum class CopyFunctionReturnType : uint8_t { CHANGED_ROWS = 0, CHANGED_ROWS_AND_FILE_LIST = 1 };
 vector<string> GetCopyFunctionReturnNames(CopyFunctionReturnType return_type);
 vector<LogicalType> GetCopyFunctionReturnLogicalTypes(CopyFunctionReturnType return_type);
+vector<LogicalType> GetTypesWithoutPartitions(const vector<LogicalType> &col_types, const vector<idx_t> &part_cols);
```
I don't think `copy_function` is the right place for these helper methods - can we move this to the `bind_copy` or `LogicalCopyToFile`?

Mytherin reviewed (commented):
Thanks for the changes! Looks good - some more comments

ykskb committed:
Make partition column skip default.

ykskb committed:
Fix python package and tests.

ykskb committed:
Merge branch 'duckdb:main' into no_partition_columns

ykskb committed:
Fixes for newly added tests.

ykskb committed:
Add implicit partition column writes.

ykskb committed:
Remove unnecessary imports.

ykskb committed:
Update Python interface with some test fixes.

ykskb commented:
Hello @Mytherin, I've made changes for PR comments:

I've made not-writing partition columns as a default behavior while adding `WRITE_PARTITION_COLUMNS` option. This change affected multiple tests on partitioned writes because they had assumed all partition data written into files. Especially the orders of columns from `select *` with Hive partitioning pattern are different because partition column data is read purely from directory names instead of duplicate column data in files. I've fixed them, and also added test cases for explicit writes of partition columns where applicable.

Also, at this point, I've made an assumption that it'll be less breaking change if we implicitly write partition columns rather than throwing an error when all the columns are specified as partition columns. I don't believe there are many use cases where all of 10 or more columns are specified as partition columns, but I thought there might be some existing use cases with less columns, looking at test cases of JSON/CSV test cases. Please let me know what you think on this. I can definitely change it back to throwing error as well.

The rest of comments on helper functions and serialization are also fixed now. Thanks!

ykskb commented:
Hi @Mytherin, just to add, the order of columns read from a data file is kept as it is read even though there are column & data found in directory names. Meanwhile, columns discovered only in directory names are appended into column list, which I believe is why they end up as last columns in results of `select *`. https://github.com/duckdb/duckdb/blob/83f552641ffa685316f988c5e32b19ee3a0f1820/src/common/multi_file_reader.cpp#L214-L225

It'll be great if you can check my comments above. Thanks very much.

ykskb committed:
Merge branch 'main' into no_partition_columns

ykskb committed:
Change implicit partition write to error.

ykskb committed:
Merge branch 'main' into no_partition_columns

ykskb committed:
Fix some tests.

ykskb committed:
Fix more tests.

ykskb committed:
Comment out copy-from of partitioned data.

ykskb committed:
Fix CSV import in partitioned write test.

ykskb commented:
Hi @Mytherin , I had another look and found those seemingly existing use cases were just for testing purposes, so I've reverted the change to throw an error when all columns are specified as partitioning columns. From my end, it really seems writing duplicate columns in files is not desired among other systems as people on the issue stated. 

I've merged the master and fixed some more tests as well. It'll be great if you can take a look when you can.

Mytherin commented:
Thanks! Agreed I think that changing the order of the columns when reading is not so critical.
